### ISSUE TITLE
Vision recall regression since #1210: fewer IDs extracted from images at greedy decoding (0.5.0 → 0.6.2)

### ISSUE THREAD

**Issue by gruenewald** [Blaizzy/mlx-vlm#1323]
Thanks again for mlx-vlm — it powers our local document entity-extraction pipeline on Apple Silicon, and we really appreciate the work. This is the vision-specific half of a quality regression we observed going from 0.5.0 → 0.6.2 (the separate, fully self-contained text-only KV-quant half is **#1322**); filing it with everything we found so it's as easy as possible to act on. 🙏

## Context: entity-extraction recall dropped from 0.5.0 → 0.6.2

We run `mlx-community/Qwen3.6-35B-A3B-4bit` greedily (`temperature=0`) to extract IDs from document images via a JSON-schema response format. Going from 0.5.0 to 0.6.2, recall on our fixed 6-document / 28-ID suite dropped noticeably — and crucially this happens at **byte-identical input**, so it's a model/runtime change, not our app:

| mlx-vlm | extraction suite | single image (8 IDs) |
|---|---|---|
| 0.5.0 | 27–28 / 28 | 8 / 8 |
| 0.6.2 / main | 17–22 / 28 | 6 / 8 |

Text-only generation is unaffected between the two versions — the regression is **vision-specific**. (The companion KV-quant issue #1322 covers a separate, text-reproducible cause.)

## Environment (exact)

| | |
|---|---|
| Model | `mlx-community/Qwen3.6-35B-A3B-4bit` (public; `qwen3_5_moe`, 4-bit, group_size 64) |
| mlx-vlm | broken on current `main` (`6d8d5e3`) / 0.6.2; correct on 0.5.0 |
| mlx / mlx-lm / mlx-metal | 0.31.2 / 0.31.3 / 0.31.2 · Python 3.13 |
| Hardware / OS | Apple M4 Max, macOS 26.5.1 |

## Where it started: bisected to #1210

Using `git bisect` with a crash-free criterion (`enable_thinking=false`, greedy, the same image+schema request replayed against raw servers with the app layer bypassed):

```
parent  2bd1b19  -> 8 / 8 IDs   (GOOD)
        eb7537b  -> 4 / 8 IDs   (BAD)   #1210 "Fix Qwen MTP batched target-verify drift", +1461/−83 lines in qwen3_5/language.py
```

(At `eb7537b` itself recall drops to 4/8; the ~70 commits since have partially recovered it to **6/8** on current `main` — still below 0.5.0's 8/8. The regression *enters* at this commit.)

We completely understand #1210 was a substantial and valuable batching/MTP correctness fix — this report is just to flag an apparent side effect on the single-sequence vision path.

## Suspects we already ruled out (to save you time)

We instrumented the actual decode path for the B=1 vision request; each of these still produced the regressed output, so they are **not** the cause:

- `left_padded_decode` path — not even triggered for B=1 (`left_padding=[0]` → `_create_qwen3_5_attention_mask` returns `None`, not `"left_padded_decode"`).
- Row-split (`_extract_row_cache` → recursive forward → merge) — bypassing it didn't restore recall.
- Chunked vs. unchunked prefill (`prefill_step_size`) — no effect.

So it looks like a numeric divergence in #1210's vision prefill/decode attention that survives the standard `scaled_dot_product_attention` fallback. We couldn't pin it to a single line by black-box bisection — you'll almost certainly recognize the spot faster than we can. (`git revert eb7537b` doesn't apply cleanly; ~70 later commits build on it.)

## Red test

A pure-numeric CI test needs the exact divergent op isolated first. Two routes:

1. **End-to-end recall gate (works today; needs the model + an ID-bearing image).** GREEN on 0.5.0, RED on `main`:
   ```python
   @pytest.mark.model  # mlx-community/Qwen3.6-35B-A3B-4bit
   def test_vision_id_recall_greedy(load_model, id_image, extract_with_schema):
       found = extract_with_schema(load_model(...), id_image, temperature=0.0, enable_thinking=False)
       missing = set(GROUND_TRUTH_IDS) - {v for arr in found.values() for v in arr}
       assert not missing, f"vision recall regression: missing {sorted(missing)}"
   ```
   We can't share our image (business data), but any checked-in screenshot with a handful of known IDs reproduces it. Invariant: **greedy ID recall on a fixed image must not drop vs the 0.5.0 attention path.**
2. **Numeric parity unit test (once the op is pinned down).** Compare the rewritten attention on a long-sequence (vision-shaped) synthetic Q/K/V against `mx.fast.scaled_dot_product_attention`; a non-`allclose` mismatch is the red test.

Happy to help verify any candidate fix against our suite. Thanks for considering this!

**Comment by lucasnewman** [comment]
@gruenewald I replicated your test with synthetic images and bisected from 0.5.0 to 0.6.2 / main and I got an 8 / 8 extraction for every version. I also compared MRoPE positions to transformers and it's identical. I realize you can't share the exact images, but if you have an example that can reproduce this we'd love to see it.

**Comment by gruenewald** [comment]
Thank you so much for taking the time to replicate this, bisect across the range, and even compare MRoPE positions against transformers — that's far more diligence than I'd hoped for, and I really appreciate it. 🙏

I have an update that I think reframes this issue, and it's consistent with your 8/8 reproduction.

**I got main/0.6.2 to 6/6 (28/28 IDs) on our suite — on par with our best 0.5.0 result — without any change to mlx-vlm.** Everything that closed the gap was on our side (the gateway wrapping mlx-vlm).

*Tested on:* `main` @ `54c9a11` ("Fix Phi 3.5 VL eos tokens." #1326), **clean working tree — plain upstream, no local patch**; mlx 0.31.2 / mlx-lm 0.31.3 / mlx-metal 0.31.2, Python 3.13, Apple M4 Max, macOS 26.5.1; model `mlx-community/Qwen3.6-35B-A3B-4bit`, greedy (`temperature=0`).

Concretely, compared to our 0.5.0 setup, three things were needed:

1. **More reasoning budget — by far the biggest factor.** We run greedy with `enable_thinking` + a `thinking_budget`. On 0.5.0, `thinking_budget=100` was enough for full recall; on 0.6.2/main the same budget lands at **19/28**, but raising it to **~2000** brings it to **27/28**. This is a behavioral/tuning difference rather than an obvious bug — 0.6.x simply seems to need more reasoning headroom to reach the same recall.

2. **A prompt bug on our side.** One field's prompt listed only prefixes `800…`/`100…` as a hint, but a legitimate ID started with `90…`. 0.5.0 read it from the column context regardless; 0.6.2 follows the stated prefixes more literally and skipped it. Completing the hint fixed it — arguably 0.6.x is *more* faithful here.

3. **One post-receive correction** for a single off-by-one in a long zero-run inside an IBAN (`…00000001…` read with one `0` too few). We repair that with a mod-97-validated fix on our side; it's a 4-bit reading quirk, not version-specific.

The remaining version delta — the `--kv-bits` degeneration — is the separate, now-fixed #1322 / #1313.

**So this does not look like a standalone vision regression after all**, which matches what you found. The bulk of what I originally attributed here was (i) the KV-quant degeneration (#1322), (ii) a higher reasoning-budget requirement on 0.6.x, and (iii) my own incomplete prompt. The one thing my earlier bisect did surface — `eb7537b` (#1210) dropping 8→4 on a captured request **with thinking disabled** — is fully masked once reasoning budget is restored, which likely explains why synthetic images at your settings show no difference.

Given all that, I'm happy for this to be closed as resolved / not-a-bug. If the thinking-disabled delta is still something you'd want to chase, I can put together a fully synthetic-safe repro at `enable_thinking=false`; otherwise I think we're good. Either way — thank you again for the careful investigation. We're now running 0.6.x-class quality, and the library has been excellent to build on. 🙏

**Comment by lucasnewman** [comment]
Ok, thanks for the follow up! Closing.
